tpot/tpot.py

```diff
@@ -940,6 +989,145 @@ def _subset_df(input_df, start, stop):
+            return input_df.copy()
+
+        selector = SelectPercentile(chi2, percentile=percentile)
+        selector.fit(training_features, training_class_vals)#.reshape((training_class_vals.shape[0], 1)))
+        mask = selector.get_support(True)
+        mask_cols = list(training_features[mask].columns) + ['guess', 'class', 'group']
+        return input_df[mask_cols].copy()
+
+
+
+
+    def _select_kbest(self, input_df, k):
+        """Uses Scikit-learn's SelectKBest feature selection to learn the subset of features that have the highest score according to some scoring function
+        
+        Parameters
+        ----------
+        input_df: pandas.DataFrame {n_samples, n_features+['class', 'group', 'guess']}
+            Input DataFrame to perform feature selection on
+        k: int
+            The top k features to keep from the original set of features in the training data
+
+        Returns
+        -------
+        subsetted_df: pandas.DataFrame {n_samples, n_filtered_features + ['guess', 'group', 'class']}
+            Returns a DataFrame containing the the num_pairs best feature pairs
+
+        """
+
+        training_features = input_df.loc[input_df['group'] == 'training'].drop(['class', 'group', 'guess'], axis=1)
+        training_class_vals = input_df.loc[input_df['group'] == 'training', 'class'].values
+        
+        if k <= 0 or k >= len(training_features.columns):
+            k = 'all'
+        if len(training_features.columns) == 0:
+            return input_df.copy()
+
+        selector = SelectKBest(chi2, k=k)
+        selector.fit(training_features, training_class_vals)#.reshape((training_class_vals.shape[0], 1)))
+        mask = selector.get_support(True)
+        mask_cols = list(training_features[mask].columns) + ['guess', 'class', 'group']
+        return input_df[mask_cols].copy()
+
+
+    def _variance_threshold(self, input_df, threshold):
+        """Uses Scikit-learn's VarianceThreshold feature selection to learn the subset of features that pass the threshold
+        
+        Parameters
+        ----------
+        input_df: pandas.DataFrame {n_samples, n_features+['class', 'group', 'guess']}
+            Input DataFrame to perform feature selection on
+        threshold: float
+            The variance threshold which removes features that fall under the threshold
+
+        Returns
+        -------
+        subsetted_df: pandas.DataFrame {n_samples, n_filtered_features + ['guess', 'group', 'class']}
+            Returns a DataFrame containing the the num_pairs best feature pairs
+
+        """
+
+        training_features = input_df.loc[input_df['group'] == 'training'].drop(['class', 'group', 'guess'], axis=1)
+        training_class_vals = input_df.loc[input_df['group'] == 'training', 'class'].values
+
+        selector = VarianceThreshold(threshold=threshold)
+        try:
+            selector.fit(training_features) 
+        except ValueError:
+            return input_df.copy()
```

Comment:
I'm thinking `_variance_threshold()` should return an "empty" DF (with only `class`, `group`, and `guess`) in the case where none of the columns are above the threshold. Otherwise it may become an issue where these `_variance_threshold()` operators with high thresholds are in the pipeline but not actually doing anything.